lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -241,23 +219,124 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
                            });
   }
 
+  void removeOp(Operation *op) { opsToRemove.push_back(op); }
+
+  Optional<size_t> getRemoteRefSend(Value val, bool mayNotExist = true) {
+    auto iter = dataflowAt.find(val);
+    if (iter != dataflowAt.end())
+      if (mayNotExist || !refSendList[iter->getSecond()].empty())
+        return iter->getSecond();
+    if (!mayNotExist) {
+      // The referenced module must have already been analyzed, error out if the
+      // dataflow at the child module is not resolved.
+      if (BlockArgument arg = val.dyn_cast<BlockArgument>())
+        arg.getOwner()->getParentOp()->emitError(
+            "reference dataflow cannot be traced back to the remote read op "
+            "for module port '")
+            << dyn_cast<FModuleOp>(arg.getOwner()->getParentOp())
+                   .getPortName(arg.getArgNumber())
+            << "'";
+      else
+        val.getDefiningOp()->emitOpError(
+            "reference dataflow cannot be traced back to the remote read op");
+      signalPassFailure();
+      return None;
+    }
+    // The Reaching def at this value is not yet known. So create an empty entry
+    // and record it.
+    return addReachingSendsEntry(val, {});
+  }
+
+  void mergeDataflow(Value src, Value dst) {
+    auto flowAtSrc = dataflowAt.find(src);
+    // If now flow at source, then just copy the info from dst. The dst might
+    // also not exist yet, in that case a new entry will be created.
+    if (flowAtSrc == dataflowAt.end())
+      dataflowAt[src] = getRemoteRefSend(dst).value();
+    else {
+      // dataflow at src exists, so merge it if an entry already exists at dest.
+      auto flowAtDst = dataflowAt.find(dst);
+      if (flowAtDst == dataflowAt.end())
+        dataflowAt[dst] = flowAtSrc->getSecond();
+      else {
+        auto oldEntry = flowAtDst->getSecond();
+        if (!refSendList[oldEntry].empty()) {
+          // If a valid dataflow already reaches the destination, emit remark.
+          // This happens when multiple connects attached to an input RefType
+          // port, or multiple instantiation (which is already an error).
+          dst.getDefiningOp()->emitRemark("has multipl reaching definitions");
+        }
+        // Dest already has an entry, make sure to update all the entries in the
+        // dataflow map to point to this new entry.
+        for (auto entr : dataflowAt) {
```

Comment:
This function makes me wonder if this would be better served by EquivalenceClasses + map from leader, or so? Haven't grok'd this all yet so might not make sense.

Reply:
You are right, this should be very well suited to using  Equivalence class, we can switch to using it.